presto-prometheus/src/main/java/com/facebook/presto/plugin/prometheus/PrometheusRecordCursor.java

```diff
@@ -0,0 +1,302 @@
+                    || type instanceof VarcharType) {
+                TypeUtils.writeNativeValue(type, builder, obj);
+            }
+        }
+    }
+
+    private static Object readObject(Type type, Block block, int position)
+    {
+        if (type instanceof ArrayType) {
+            Type elementType = ((ArrayType) type).getElementType();
+            return getArrayFromBlock(elementType, block.getBlock(position));
+        }
+        else if (type instanceof MapType) {
+            return getMapFromBlock(type, block.getBlock(position));
+        }
+        else {
+            if (type.getJavaType() == Slice.class) {
+                Slice slice = (Slice) requireNonNull(TypeUtils.readNativeValue(type, block, position));
+                return (type instanceof VarcharType) ? slice.toStringUtf8() : slice.getBytes();
+            }
+
+            return TypeUtils.readNativeValue(type, block, position);
+        }
+    }
+
+    private static List<Object> getArrayFromBlock(Type elementType, Block block)
+    {
+        ImmutableList.Builder<Object> arrayBuilder = ImmutableList.builder();
+        for (int i = 0; i < block.getPositionCount(); ++i) {
+            arrayBuilder.add(readObject(elementType, block, i));
+        }
+        return arrayBuilder.build();
+    }
+
+    private static Map<String, String> metricHeaderToMap(Map<String, String> mapToConvert)
+    {
+        return ImmutableMap.<String, String>builder().putAll(mapToConvert).build();
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return totalBytes;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return 0;
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        checkArgument(field < columnHandles.size(), "Invalid field index");
+        return columnHandles.get(field).getColumnType();
+    }
+
+    @Override
+    public boolean advanceNextPosition()
+    {
+        if (!metricsItr.hasNext()) {
+            return false;
+        }
+        fields = metricsItr.next();
+        return true;
+    }
+
+    @Override
+    public boolean getBoolean(int field)
+    {
+        return true;
+    }
+
+    @Override
+    public long getLong(int field)
+    {
+        Type type = getType(field);
+        if (type.equals(TIMESTAMP_COLUMN_TYPE)) {
+            Instant dateTime = (Instant) requireNonNull(getFieldValue(field));
+            // render with the fixed offset of the Presto server
+            int offsetMinutes = dateTime.atZone(ZoneId.systemDefault()).getOffset().getTotalSeconds() / 60;
+            return packDateTimeWithZone(dateTime.toEpochMilli(), offsetMinutes);
+        }
+        else {
+            throw new PrestoException(NOT_SUPPORTED, "Unsupported type " + getType(field));
+        }
+    }
+
+    @Override
+    public double getDouble(int field)
+    {
+        checkFieldType(field, DOUBLE);
+        return (double) requireNonNull(getFieldValue(field));
```

Comment:
have requireNonNull in the return value is not good. Let's add checks or throw exception in getFieldValue

Reply:
ok